### PR TITLE
Added MonoBank payment system to GreenCityUBS

### PR DESCRIPTION
**Pull request by KizerovDmitriy**
# GreenCityUBS PR
## Summary Of Changes 🔥

 - Added new DTOs for monobank API.
 - Added new endpoint to validate paymet.
 - Added logic to form payment request to API.
 - Added tests.

## Issue link

[#7508](https://github.com/ita-social-projects/GreenCity/issues/7508)

# PR Checklist Forms

_(to be filled out by PR submitter)_
- [x] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells or duplication
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 1]
```diff
@@ -581,11 +601,64 @@ public WayForPayOrderResponse saveFullOrderToDB(OrderResponseDto dto, String uui
...
+        return getPaymentRequestDto(order, checkoutResponse.getPageUrl());
+    }
+
+    private MonoBankPaymentRequestDto formPaymentRequestForMonoBank(Long orderId, long sumToPayInCoins, User user) {
+        Order order = orderRepository.findById(orderId)
+            .orElseThrow(() -> new NotFoundException(ORDER_WITH_CURRENT_ID_DOES_NOT_EXIST));
+
+        return MonoBankPaymentRequestDto.builder()
+            .amount(1)
+            // .amount((int) sumToPayInCoins)
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:637]: do we need this comment?

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 2]
```diff
@@ -2004,4 +2081,72 @@ public List<LocationsDto> getAllLocationsByCourierId(Long courierId) {
...
+            .paymentSystem(response.getPaymentInfo().getPaymentSystem())
+            .senderEmail(order.getUser().getRecipientEmail())
+            .paymentId(response.getInvoiceId())
+            .order(order)
+            .paymentType(PaymentType.AUTO)
+            .paymentStatus(PaymentStatus.UNPAID)
+            .build();
+    }
+
+    private void checkPaymentStatus(MonoBankPaymentResponseDto response, Payment payment, Order order) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: this method does not check the status and should be renamed according to the its logic

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 3]
```diff
@@ -2004,4 +2081,72 @@ public List<LocationsDto> getAllLocationsByCourierId(Long courierId) {
...
+        if (MonoBankStatuses.FAILURE.getName().equalsIgnoreCase(response.getStatus())
+            || MonoBankStatuses.REVERSED.getName().equalsIgnoreCase(response.getStatus())) {
+            payment.setPaymentStatus(PaymentStatus.UNPAID);
+            order.setOrderPaymentStatus(OrderPaymentStatus.UNPAID);
+            paymentRepository.save(payment);
+            orderRepository.save(order);
+        }
+    }
+
+    private String convertToDateOnly(String isoDateString) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: you can try
```
OffsetDateTime.parse(isoDateString).format(DateTimeFormatter.ISO_LOCAL_DATE);
```

File: core/src/main/java/greencity/controller/OrderController.java [hunk 4]
```diff
@@ -530,4 +531,25 @@ public ResponseEntity<List<LocationsDto>> getAllLocationsByCourierId(
...
+     * @param response the payment response received from Monobank, containing
+     *                 details such as transaction ID, status, and amount.
+     */
+    @Operation(summary = "Receive payment information from Monobank API")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = HttpStatuses.OK),
+        @ApiResponse(responseCode = "400", description = HttpStatuses.BAD_REQUEST)
+    })
+    @PostMapping("/monobank/payments")
+    public void receivePaymentFromMonoBank(@RequestBody MonoBankPaymentResponseDto response) {
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/OrderController.java:551]: based on the method name if you call it you expect to `receive` something. Why return type is `void`?

File: core/src/main/java/greencity/controller/OrderController.java [hunk 5]
```diff
@@ -530,4 +531,25 @@ public ResponseEntity<List<LocationsDto>> getAllLocationsByCourierId(
...
+        log.info("Response from MONOBANK API: {}", response);
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/OrderController.java:552]: do you need to log the whole request?
Please check if it does not contain sensitive info

File: core/src/main/java/greencity/controller/OrderController.java (hunk 4 above)
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/OrderController.java:551]: Why the name and type of input param is Response not Request?

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 6]
```diff
@@ -2004,4 +2081,72 @@ public List<LocationsDto> getAllLocationsByCourierId(Long courierId) {
...
+        if (MonoBankStatuses.SUCCESS.getName().equalsIgnoreCase(response.getStatus())) {
+            payment.setPaymentStatus(PaymentStatus.PAID);
+            order.setOrderPaymentStatus(OrderPaymentStatus.PAID);
+            paymentRepository.save(payment);
+            orderRepository.save(order);
+            eventService.save(OrderHistory.ORDER_PAID, OrderHistory.SYSTEM, order);
+            eventService.save(OrderHistory.ADD_PAYMENT_SYSTEM + payment.getPaymentId(),
+                OrderHistory.SYSTEM, order);
+        }
+        if (MonoBankStatuses.FAILURE.getName().equalsIgnoreCase(response.getStatus())
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: what should we do if we have some other status (not `SUCCESS`, `FAILURE` or `REVERSED`)?

**Comment by RomaMocherniuk** [comment]
please fix this Sonare issue
image

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java (hunk 1 above)
Comment (KizerovDmitriy) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:637]: This is a temporary comment, for payment testing. Setting amount of payment to 1 koop

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java (hunk 3 above)
Comment (KizerovDmitriy) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: thx for the advice

File: core/src/main/java/greencity/controller/OrderController.java (hunk 4 above)
Comment (KizerovDmitriy) [reply core/src/main/java/greencity/controller/OrderController.java:551]: This method receives a response from the MonoBank API, checks the response in the current endpoint, and validates the payment if the payment status is successful. No responses need to be sent back to the API.

File: core/src/main/java/greencity/controller/OrderController.java (hunk 5 above)
Comment (KizerovDmitriy) [reply core/src/main/java/greencity/controller/OrderController.java:552]: I did not have the opportunity to test this case locally, so I added temporary logging to verify its correctness. If it works correctly, I will remove the logging.

File: core/src/main/java/greencity/controller/OrderController.java (hunk 4 above)
Comment (KizerovDmitriy) [reply core/src/main/java/greencity/controller/OrderController.java:551]: This endpoint receive response from monobank API

**Comment by KizerovDmitriy** [comment]
> please fix this Sonare issue image

This is because of the commented line where we set the payment:
image

**Review by HelenSotnik — APPROVED** [review]

**Review by lukovskyj — APPROVED** [review]

**Review by K4sik — APPROVED** [review]

File: core/src/main/java/greencity/controller/OrderController.java (hunk 4 above)
Comment (RomaMocherniuk) [reply core/src/main/java/greencity/controller/OrderController.java:551]: `validates the payment if the payment status is successful`
What does happen if the response is not successful?

File: service-api/src/main/java/greencity/dto/order/PaymentSystemResponse.java [hunk 7]
```diff
@@ -13,7 +13,7 @@
 @Setter
 @EqualsAndHashCode
 @Builder
-public class WayForPayOrderResponse {
+public class PaymentSystemResponse {
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/order/PaymentSystemResponse.java]: We can use `record` here

File: service-api/src/main/java/greencity/dto/payment/PaymentWayForPayRequestDto.java [hunk 8]
```diff
@@ -15,7 +15,7 @@
 @AllArgsConstructor
 @Builder
 @ToString
-public class PaymentRequestDto {
+public class PaymentWayForPayRequestDto {
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/payment/PaymentWayForPayRequestDto.java:18]: the same here

File: service-api/src/main/java/greencity/dto/payment/monobank/PaymentInfo.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+@Builder
+public class PaymentInfo {
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/payment/monobank/PaymentInfo.java]: please use `record` for all classes that contain data only

**Comment by RomaMocherniuk** [comment]
Please fix sonar issue

**Pull request merged**